File: ReconstructItinerary.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<set>

using namespace std;

class Solution {
public:
    vector<string> findItinerary(vector<pair<string, string> > tickets) {
        
        unordered_map<string, multiset<string> > m;
        
        for(auto a : tickets) {
        	m[a.first].insert(a.second);
		}
		
		vector<string> result;
		
		dfs(m, "JFK", result);
		
		return vector<string>(result.rbegin(), result.rend());
        
    }
    
    void dfs(unordered_map<string, multiset<string> >& m, string node, vector<string>& result) {
    	
    	while(!m[node].empty()) {
    		string t = *m[node].begin();
    		m[node].erase(m[node].begin());
    		dfs(m, t, result);
		}
		
		result.push_back(node);
    	
	}
    
};

int main() {
	
	int n;
	cin>>n;
	
	vector<pair<string, string> > tickets;
	for(int i = 0; i < n; i++) {
		string from, to;
		cin>>from>>to;
		tickets.push_back(make_pair(from, to));
	}
	
	Solution *solution = new Solution();
	vector<string> result = solution->findItinerary(tickets);
	
	for(int i = 0; i < result.size(); i++) cout<<result[i]<<"  ";
	
	return 0;
	
}

File: ReconstructItinerary.py
class Solution:
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        def dfs(pg):
            while pg in m and m[pg]:
                t = m[pg][0]
                m[pg].pop(0)
                dfs(t)
            result.append(pg)
        m = {}
        for tick in tickets:
            if tick[0] in m:
                m[tick[0]].append(tick[1])
            else:
                m[tick[0]] = [tick[1]]
        for k in m.keys():
            m[k].sort()

        result = []
        dfs('JFK')
        result.reverse()
        return result


if __name__ == '__main__':
    tickets = []
    n = int(input())
    for _ in range(n):
        tickets.append([a for a in input().split()])
    solution = Solution()
    print(solution.findItinerary(tickets))
